sdk/py/apepay/streams.py

```diff
@@ -20,139 +17,94 @@
+        Funding rate, in tokens per second, of Stream in human-readable decimal form.
         """
         return Decimal(self.amount_per_second) / Decimal(10 ** self.token.decimals())
 
-    def estimate_funding(self, period: timedelta) -> int:
+    def estimate_funding(self, period: timedelta) -> Decimal:
         """
-        Useful for estimating how many tokens you need to add to extend for a specific time period.
+        Useful for displaying how many tokens you need to add to extend for a specific time period.
         """
-        return int(period.total_seconds() * self.amount_per_second)
+        return int(period.total_seconds()) * self.funding_rate
 
     @cached_property
     def start_time(self) -> datetime:
+        # NOTE: This cannot be updated
         return datetime.fromtimestamp(self.info.start_time)
 
     @cached_property
-    def reason(self) -> HexBytes | str | dict:
-        try:
-            reason_str = self.info.reason.decode("utf-8")
+    def products(self) -> list[HexBytes]:
+        # NOTE: This cannot be updated
+        return self.info.products
 
-        except Exception:
-            return self.info.reason
-
-        try:
-            return json.loads(reason_str)
-
-        except (Exception, json.JSONDecodeError):
-            return reason_str
+    @property
+    def owner(self) -> AddressType:
+        return self.info.owner
 
     @property
     def last_pull(self) -> datetime:
         return datetime.fromtimestamp(self.info.last_pull)
 
     @property
-    def amount_unlocked(self) -> int:
-        return self.contract.amount_unlocked(self.creator, self.stream_id)
+    def amount_claimable(self) -> int:
+        return self.contract.amount_claimable(self.id)
 
     @property
-    def amount_locked(self) -> int:
-        return self.info.funded_amount - self.amount_unlocked
+    def amount_refundable(self) -> int:
+        # NOTE: Max `.amount_claimable` can be is `.funded_amount`
+        return self.info.funded_amount - self.amount_claimable
 
     @property
     def time_left(self) -> timedelta:
-        seconds = self.contract.time_left(self.creator, self.stream_id)
-        return timedelta(seconds=min(MAX_DURATION_SECONDS, seconds))
+        seconds = self.contract.time_left(self.id)
+        assert seconds < MAX_DURATION_SECONDS, "Invaraint wrong"
+        return timedelta(seconds=seconds)
 
     @property
     def total_time(self) -> timedelta:
-        info = self.info  # NOTE: Avoid calling contract twice
-        # NOTE: Measure time-duration of unclaimed amount remaining (locked and unlocked)
-        max_life = int(info.funded_amount / info.amount_per_second)
+        info = self.info  # NOTE: Avoid calling contract twice by caching
+
+        # NOTE: Measure time-duration of unclaimed amount remaining
+        remaining_life = int(info.funded_amount / info.amount_per_second)
+        assert remaining_life < MAX_DURATION_SECONDS, "Invariant wrong"
```

Comment:
may want an exception here

Reply:
This is a usage invariant that should never be violated, but I wanted it here for when you do testing to make sure it raises properly (`remaining_life` is time in seconds, violating this invariant means you want a stream to run for the rest of the life of the universe)

Basically only if your streaming rates are very small and your funding rates are very large would it be a problem

Reply:
Ah ok. Got it. However, my experience working on the framework -- if there is a hacky bad thing people can do, they will somehow do it.

Reply:
It will raise a weirder error about C types being out of bounds